consensus/src/da_replica.rs

```diff
@@ -0,0 +1,526 @@
+    }
+
+    /// Replica task for sequencing consensus that spins until a vote can be made or timeout is
+    /// hit.
+    ///
+    /// Returns the new leaf if it's valid.
+    #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Sequencing Replica Task", level = "error")]
+    #[allow(clippy::type_complexity)]
+    async fn find_valid_msg<'a>(
+        &self,
+        view_leader_key: TYPES::SignatureKey,
+        consensus: RwLockUpgradableReadGuard<'a, Consensus<TYPES, DALeaf<TYPES>>>,
+    ) -> (
+        RwLockUpgradableReadGuard<'a, Consensus<TYPES, DALeaf<TYPES>>>,
+        Option<DALeaf<TYPES>>,
+    ) {
+        let lock = self.proposal_collection_chan.lock().await;
+        let mut invalid_qc = false;
+        let leaf = loop {
+            let msg = lock.recv().await;
+            info!("recv-ed message {:?}", msg.clone());
+            if let Ok(msg) = msg {
+                // stale/newer view messages should never reach this specific task's receive channel
+                if Into::<ConsensusMessage<_, _, _>>::into(msg.clone()).view_number()
+                    != self.cur_view
+                {
+                    continue;
+                }
+                match msg {
+                    ProcessedConsensusMessage::Proposal(p, sender) => {
+                        if view_leader_key != sender {
+                            continue;
+                        }
+
+                        let mut valid_leaf = None;
+                        let vote_token = self.api.make_vote_token(self.cur_view);
+                        match vote_token {
+                            Err(e) => {
+                                error!(
+                                    "Failed to generate vote token for {:?} {:?}",
+                                    self.cur_view, e
+                                );
+                            }
+                            Ok(None) => {
+                                info!("We were not chosen for committee on {:?}", self.cur_view);
+                            }
+                            Ok(Some(vote_token)) => {
+                                info!("We were chosen for committee on {:?}", self.cur_view);
+
+                                let message;
+
+                                // Construct the leaf.
+                                let justify_qc = p.data.justify_qc;
+                                let parent_commitment = match self.parent_leaf().await {
+                                    Some(parent) => parent.commit(),
+                                    None => {
+                                        break None;
+                                    }
+                                };
+                                let block_commitment = p.data.block_commitment;
+                                let leaf = DALeaf {
+                                    view_number: self.cur_view,
+                                    height: 0,
```

Comment:
Should this depend on the parent's height?

Reply:
Oh we should add height to `CommitmentProposal` it's on the leaf and also `ValidatingProposal` and the leader already depends on parent height to set this value

Reply:
I was thinking about using `parent.height` but it's not guaranteed to be the same as the height the leader used, so yeah, I think it's a good idea to add `height` to `CommitmentProposal`.